076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,389 @@
+ - Strimzi cannot guarantee that the issuer will issue every certificate signed by the same root CA certificate. 
+   - For example cert-manager will if possible include the CA certificate to trust when it issues the certificate (if the CA certificate is not a public one), however this is per certificate and not always provided.
+   - Using those solutions therefore requires TLS peers to be able to trust multiple root CA certificates.
+ - These solutions cannot be responsible for rolling updates to components like Kafka.
+   - This means that end-entity certificate generation and the rollout of trust in the CA certificates referenced in those end-entity certificates, needs to be decoupled.
+ - These solutions might take a varying amount of time to issue a certificate.
+   - Since we do not know how long certificate issuance might take, we can't block the reconcile loop waiting for certificates to be issued.
+   - This means it can't be assumed that e.g. the CA certificate is already present when the component reconcilers run (since it not be issued yet).
+This would be highly valuable for organizations with compliance requirements with regard to certificates.
+
+Currently, it is hard to integrate one of these certificate management solutions for a couple of reasons:
+ - The state of certificates, particularly CA certificates is determined based on multiple different annotations and fields on internal Java objects.
+   This makes it hard to comprehend the current state, particularly when doing a renewal and tricky to reason about edge cases that might occur.
+ - The component reconcilers directly interact with the Java class that represents the Cluster CA.
+   For example, this class contains methods not only to manage the CA certificate and private key, but also generate certificates for Strimzi components (like Cruise Control) and store state about renewals that happened during the current reconcile loop.
+   This makes it hard to drop in an alternative to the existing Strimzi Cluster CA, since it's behaviour is relied upon throughout the reconcile loop.
+
+This proposal describes changes to make it easier to reason about the current state of CA certificates and track the roll out of trust.
+
+## Proposal
+
+Strimzi should be updated to decouple the management of the cluster CA certificate, and the consumption of the CA certificate by components (such as ZooKeeper, Kafka and cluster operator).
+
+To do this we will introduce two new Kubernetes Secrets.
+The first Kubernetes Secret, called <CLUSTER_NAME>-cluster-ca-trust-state (referred to from here onwards as trust-state Secret) is used to keep track of all Cluster CA certificates that are currently in use and their current state with respect to issued end entity certificates.
+The second Kubernetes Secret, called <CLUSTER_NAME>-cluster-ca-trusted-certs (referred to from here onwards as trusted-certs Secret) is used to store all Cluster CA certificates that are currently trusted by components.
```

Comment:
Why do these need to be different secrets?

Reply:
In a previous iteration I had them combined with the Secret data having the format `${fingerprint}.${state}: <PEM encoded CA certificate>` but @fvaleri noted that this could be a problem if we introduced a mechanism for truststore hot reloading.

Also it seemed better separation of concerns to me.

Reply:
I would agree that the state and the actual certificate should be separate. But that does not mean you cannot have the same secret storing both ... `<fingerprint>.crt` for the actual cert and `<fingerprint>.state` for some JSON describing the state.